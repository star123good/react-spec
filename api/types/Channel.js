// @flow
const Channel = /* GraphQL */ `
	type ChannelMembersConnection {
		pageInfo: PageInfo!
		edges: [ChannelMemberEdge!]
	}

	type ChannelMemberEdge {
		cursor: String!
		node: User!
	}

	type ChannelThreadsConnection {
		pageInfo: PageInfo!
		edges: [ChannelThreadEdge!]
	}

	type ChannelThreadEdge {
		cursor: String!
		node: Thread!
	}

	type ChannelMetaData {
		threads: Int
		members: Int
	}

	input CreateChannelInput {
		name: String!
		slug: LowercaseString!
		description: String
		communityId: ID!
		isPrivate: Boolean
		isDefault: Boolean
	}

	input EditChannelInput {
		name: String
		slug: LowercaseString
		description: String
		isPrivate: Boolean
		channelId: ID!
	}

	enum PendingActionType {
		block
		approve
	}

	input TogglePendingUserInput {
		channelId: ID!
		userId: ID!
		action: PendingActionType!
	}

	input UnblockUserInput {
		channelId: ID!
		userId: ID!
	}

	type JoinSettings {
		tokenJoinEnabled: Boolean
		token: String
	}

	type Channel {
		id: ID!
		createdAt: Date!
		modifiedAt: Date
		name: String!
		description: String!
		slug: LowercaseString!
		isPrivate: Boolean
		isDefault: Boolean
		isArchived: Boolean
    channelPermissions: ChannelPermissions! @cost(complexity: 1)
		communityPermissions: CommunityPermissions!
    community: Community! @cost(complexity: 1)
		threadConnection(first: Int = 10, after: String): ChannelThreadsConnection! @cost(complexity: 1, multiplier: "first")
    memberConnection(first: Int = 10, after: String): ChannelMembersConnection! @cost(complexity: 1, multiplier: "first")
		memberCount: Int!
    metaData: ChannelMetaData @cost(complexity: 1)
    pendingUsers: [User] @isAuthed @cost(complexity: 3)
		blockedUsers: [User] @isAuthed @cost(complexity: 3)
		moderators: [User] @cost(complexity: 3)
		owners: [User] @cost(complexity: 3)
		joinSettings: JoinSettings 
		slackSettings: ChannelSlackSettings @isAuthed
	}

	extend type Query {
    channel(id: ID, channelSlug: LowercaseString, communitySlug: LowercaseString): Channel @cost(complexity: 1)
	}

	input ArchiveChannelInput {
		channelId: ID!
	}

	input RestoreChannelInput {
		channelId: ID!
	}

	input JoinChannelWithTokenInput {
		communitySlug: LowercaseString!
		channelSlug: LowercaseString!
		token: String!
	}

	input EnableChannelTokenJoinInput {
		id: ID!
	}

	input DisableChannelTokenJoinInput {
		id: ID!
	}

	input ResetChannelJoinTokenInput {
		id: ID!
	}

	extend type Mutation {
		createChannel(input: CreateChannelInput!): Channel @isAuthed
		editChannel(input: EditChannelInput!): Channel @isAuthed
		deleteChannel(channelId: ID!): Boolean @isAuthed
		toggleChannelSubscription(channelId: ID!): Channel @isAuthed
		joinChannelWithToken(input: JoinChannelWithTokenInput!): Channel @isAuthed
		toggleChannelNotifications(channelId: ID!): Channel @isAuthed
		togglePendingUser(input: TogglePendingUserInput!): Channel @isAuthed
		unblockUser(input: UnblockUserInput!): Channel @isAuthed
		archiveChannel(input: ArchiveChannelInput!): Channel @isAuthed
		restoreChannel(input: RestoreChannelInput!): Channel @isAuthed
		enableChannelTokenJoin(input: EnableChannelTokenJoinInput!): Channel @isAuthed
		disableChannelTokenJoin(input: DisableChannelTokenJoinInput!): Channel @isAuthed
		resetChannelJoinToken(input: ResetChannelJoinTokenInput!): Channel @isAuthed
	}
`;

module.exports = Channel;
