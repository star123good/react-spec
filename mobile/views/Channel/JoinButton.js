// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { Button } from '../../components/Button';
import compose from 'recompose/compose';
import { type GetChannelType } from '../../../shared/graphql/queries/channel/getChannel';
import toggleChannelSubscription, {
  type ToggleChannelSubscriptionType,
  type ToggleChannelSubscriptionProps,
} from '../../../shared/graphql/mutations/channel/toggleChannelSubscription';
import { JoinButtonWrapper } from './style';
import MutationWrapper from '../../components/MutationWrapper';
import { addToast } from '../../actions/toasts';

type Props = {
  ...$Exact<ToggleChannelSubscriptionProps>,
  channel: GetChannelType,
  dispatch: Dispatch<Object>,
};

type State = {
  joinedDuringSession: boolean,
};

class JoinButton extends React.Component<Props, State> {
  state = {
    isLoading: false,
    joinedDuringSession: false,
  };

  onComplete = (data: ToggleChannelSubscriptionType) => {
    const joined =
      data.data.toggleChannelSubscription.channelPermissions.isMember;

    this.setState({
      joinedDuringSession: joined,
    });
  };

  onError = (err: any) => {
    return this.props.dispatch(
      addToast({
        type: 'error',
        message: err.message,
      })
    );
  };

  render() {
    const { channel } = this.props;
    const { joinedDuringSession } = this.state;

    if (joinedDuringSession) {
      return (
        <JoinButtonWrapper>
          <MutationWrapper
            mutation={this.props.toggleChannelSubscription}
            variables={{ channelId: channel.id }}
            onComplete={this.onComplete}
            onError={this.onError}
            render={({ isLoading, onPressHandler }) => (
              <Button
                onPress={onPressHandler}
                state={isLoading ? 'loading' : null}
                icon={'member-remove'}
                title={'Leave channel'}
              />
            )}
          />
        </JoinButtonWrapper>
      );
    }

    if (channel.channelPermissions.isMember) return null;

    return (
      <JoinButtonWrapper>
        <MutationWrapper
          mutation={this.props.toggleChannelSubscription}
          variables={{ channelId: channel.id }}
          onComplete={this.onComplete}
          onError={this.onError}
          render={({ isLoading, onPressHandler }) => (
            <Button
              onPress={onPressHandler}
              state={isLoading ? 'loading' : null}
              icon={'member-add'}
              title={'Join channel'}
            />
          )}
        />
      </JoinButtonWrapper>
    );
  }
}

export default compose(connect(), toggleChannelSubscription)(JoinButton);
