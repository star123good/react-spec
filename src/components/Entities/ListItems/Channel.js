// @flow
import * as React from 'react';
import { Link } from 'react-router-dom';
import Icon from 'src/components/icons';
import { Row, Content, Label, Description, Actions } from './style';

type Props = {
  channelObject: Object,
  id: string,
  name?: string,
  description?: ?string,
  children?: React.Node,
};

export const ChannelListItem = (props: Props) => {
  const { channelObject, name, description, children } = props;

  return (
    <Link to={`/${channelObject.community.slug}/${channelObject.slug}`}>
      <Row>
        <Content>
          {name && (
            <Label>
              {channelObject.isPrivate && (
                <Icon glyph="private-outline" size={14} />
              )}

              {name}
            </Label>
          )}

          {description && <Description>{description}</Description>}
        </Content>

        <Actions>
          <Icon glyph="view-forward" size={24} />

          {children}
        </Actions>
      </Row>
    </Link>
  );
};

export default ChannelListItem;
