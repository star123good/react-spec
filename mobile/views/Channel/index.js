// @flow
import React, { Component } from 'react';
import compose from 'recompose/compose';
import {
  getChannelById,
  type GetChannelType,
} from '../../../shared/graphql/queries/channel/getChannel';
import getChannelThreadConnection from '../../../shared/graphql/queries/channel/getChannelThreadConnection';
import ViewNetworkHandler from '../../components/ViewNetworkHandler';
import ThreadFeed from '../../components/ThreadFeed';
import Loading from '../../components/Loading';
import JoinButton from './JoinButton';
import { track, transformations, events } from '../../utils/analytics';

import {
  Wrapper,
  CoverPhoto,
  CoverPhotoFill,
  CoverPhotoContainer,
  ProfilePhoto,
  ProfilePhotoContainer,
  ProfileDetailsContainer,
  Name,
  Username,
  Description,
  ThreadFeedDivider,
} from './style';
import ErrorBoundary from '../../components/ErrorBoundary';
import { FullscreenNullState } from '../../components/NullStates';
import type { NavigationProps } from 'react-navigation';

type Props = {
  isLoading: boolean,
  hasError: boolean,
  navigation: NavigationProps,
  data: {
    channel?: GetChannelType,
  },
};

const ChannelThreadFeed = compose(getChannelThreadConnection)(ThreadFeed);

class Channel extends Component<Props> {
  trackView = () => {
    const { data: { channel } } = this.props;
    if (!channel) return;
    track(events.CHANNEL_VIEWED, {
      channel: transformations.analyticsChannel(channel),
      community: transformations.analyticsCommunity(channel.community),
    });
  };

  setTitle = () => {
    const { data: { channel }, navigation } = this.props;
    let title;
    if (channel) {
      title = channel.name;
    } else {
      title = 'Loading channel...';
    }
    const oldTitle = navigation.getParam('title', null);
    if (oldTitle && oldTitle === title) return;
    navigation.setParams({ title });
  };

  componentDidMount() {
    this.trackView();
    this.setTitle();
  }

  componentDidUpdate(prev) {
    const curr = this.props;
    const first = !prev.data.channel && curr.data.channel;
    const changed =
      prev.data.channel &&
      curr.data.channel &&
      prev.data.channel.id !== curr.data.channel.id;
    if (first || changed) {
      this.trackView();
    }

    this.setTitle();
  }

  render() {
    const { data, isLoading, hasError, navigation } = this.props;

    if (data.channel) {
      const { channel } = data;

      return (
        <Wrapper>
          <ChannelThreadFeed
            navigation={navigation}
            id={channel.id}
            activeChannel={channel.id}
            activeCommunity={channel.community.id}
            ListHeaderComponent={
              <ErrorBoundary alert>
                <CoverPhotoContainer>
                  {channel.community.coverPhoto ? (
                    <CoverPhoto
                      resizeMode={'cover'}
                      source={{ uri: channel.community.coverPhoto }}
                    />
                  ) : (
                    <CoverPhotoFill />
                  )}
                </CoverPhotoContainer>

                <ProfilePhotoContainer>
                  <ProfilePhoto
                    source={{ uri: channel.community.profilePhoto }}
                  />
                </ProfilePhotoContainer>

                <ProfileDetailsContainer>
                  <Username>{channel.community.name}</Username>
                  <Name>{channel.name}</Name>
                  <Description>{channel.description}</Description>
                  <JoinButton channel={channel} />
                </ProfileDetailsContainer>

                <ThreadFeedDivider />
              </ErrorBoundary>
            }
          />
        </Wrapper>
      );
    }

    if (isLoading) {
      return (
        <Wrapper>
          <Loading />
        </Wrapper>
      );
    }

    if (hasError) {
      return <FullscreenNullState />;
    }

    return null;
  }
}

export default compose(getChannelById, ViewNetworkHandler)(Channel);
