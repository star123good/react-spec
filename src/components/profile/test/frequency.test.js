// @flow
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Channel from '../channel';

const data = {
  title: 'Test channel',
  subtitle: 'Whats up',
  description: 'Testing the channel profile',
  id: 'asdf-123',
};

it('should render with no size', () => {
  const wrapper = shallow(<Channel data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a mini size', () => {
  const wrapper = shallow(<Channel size="mini" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a small size', () => {
  const wrapper = shallow(<Channel size="small" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a medium size', () => {
  const wrapper = shallow(<Channel size="medium" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a large size', () => {
  const wrapper = shallow(<Channel size="large" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a full size', () => {
  const wrapper = shallow(<Channel size="full" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
