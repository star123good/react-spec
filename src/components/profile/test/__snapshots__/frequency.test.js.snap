exports[`test should render with a full size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }
  size="full">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the channel profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="channel" />
</pure(CardPure)>
`;

exports[`test should render with a large size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }
  size="large">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the channel profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="channel" />
</pure(CardPure)>
`;

exports[`test should render with a medium size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }
  size="medium">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the channel profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a mini size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }
  size="mini">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a small size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }
  size="small">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with no size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the channel profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test channel",
    }
  }>
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test channel
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;
